### PR TITLE
Release Planning and Roadmapping: Dashboard Metrics & Planning Analytics Implementation #159

### PR DESCRIPTION
**Pull request by codecharlan**
## Objective
Implement API endpoints and calculation logic to support release dashboards and comprehensive planning analytics.

#### 1. GET /api/releases/{code}/dashboard
Release dashboard with overview, feature breakdown, progress metrics, health indicators, and timeline data.

**Example Response:**
```json
{
  "releaseCode": "REL-2024-Q1",
  "releaseName": "Q1 2024 Release",
  "status": "IN_PROGRESS",
  "overview": {
    "totalFeatures": 45,
    "completedFeatures": 12,
    "inProgressFeatures": 18,
    "blockedFeatures": 3,
    "pendingFeatures": 12,
    "completionPercentage": 26.7,
    "estimatedDaysRemaining": 28
  },
  "healthIndicators": {
    "timelineAdherence": "DELAYED",
    "riskLevel": "MEDIUM",
    "blockedFeatures": 3
  },
  "timeline": {
    "startDate": "2024-01-15T00:00:00Z",
    "plannedEndDate": "2024-03-31T00:00:00Z",
    "estimatedEndDate": "2024-04-15T00:00:00Z",
    "actualEndDate": null
  },
  "featureBreakdown": {
    "byStatus": {
      "NEW": 5,
      "IN_PROGRESS": 3,
      "ON_HOLD": 1,
      "RELEASED": 12
    },
    "byOwner": {
      "user1@company.com": 5,
      "user2@company.com": 8,
      "unassigned": 3
    }
  }
}
```

#### 2. GET /api/releases/{code}/metrics
Detailed metrics for analytics including completion rate, velocity, blocked time, and workload distribution.

**Example Response:**
```json
{
  "releaseCode": "REL-2024-Q1",
  "releaseStatus": "IN_PROGRESS",
  "completionRate": 26.7,
  "velocity": {
    "featuresPerWeek": 3.2,
    "averageCycleTime": 8.5
  },
  "blockedTime": {
    "totalBlockedDays": 15,
    "averageBlockedDuration": 5.0,
    "percentageOfTime": 8.5,
    "blockageReasons": {}
  },
  "workloadDistribution": {
    "byOwner": [
      {
        "owner": "john.doe",
        "assignedFeatures": 8,
        "completedFeatures": 3,
        "inProgressFeatures": 4,
        "blockedFeatures": 1,
        "utilizationRate": 85.0
      }
    ]
  }
}
```

#### 1. GET /api/planning/health
Planning health report including releases by status, at-risk counts, and planning accuracy.

**Example Response:**
```json
{
  "releasesByStatus": {
    "DRAFT": 8,
    "IN_PROGRESS": 12,
    "RELEASED": 45,
    "CANCELLED": 2
  },
  "atRiskReleases": {
    "overdue": 3,
    "criticallyDelayed": 2,
    "total": 5
  },
  "planningAccuracy": {
    "onTimeDelivery": 73.2,
    "averageDelay": 12.5,
    "estimationAccuracy": 78.9
  }
}
```

#### 2. GET /api/planning/trends
Trend data for releases completed, average duration, and planning accuracy over time.

**Example Response:**
```json
{
  "releasesCompleted": {
    "trend": [
      { "period": "2023-02", "value": 4 },
      { "period": "2023-03", "value": 6 },
      { "period": "2024-01", "value": 5 }
    ],
    "total": 58
  },
  "averageReleaseDuration": {
    "trend": [
      { "period": "2023-02", "value": 89.5 },
      { "period": "2023-03", "value": 92.3 },
      { "period": "2024-01", "value": 87.1 }
    ],
    "current": 87.1
  },
  "planningAccuracyTrend": {
    "onTimeDelivery": [
      { "period": "2023-02", "value": 68.2 },
      { "period": "2023-03", "value": 71.5 },
      { "period": "2024-01", "value": 73.2 }
    ]
  }
}
```

#### 3. GET /api/planning/capacity
Capacity planning data including workload distribution, commitments, and overallocation warnings.

**Example Response:**
```json
{
  "overallCapacity": {
    "totalResources": 45,
    "utilizationRate": 87.3,
    "availableCapacity": 12.7,
    "overallocatedResources": 3
  },
  "workloadByOwner": [
    {
      "owner": "john.doe",
      "currentWorkload": 8,
      "capacity": 10,
      "utilizationRate": 80.0,
      "futureCommitments": 12,
      "overallocationRisk": "NONE"
    },
    {
      "owner": "jane.smith",
      "currentWorkload": 12,
      "capacity": 10,
      "utilizationRate": 120.0,
      "futureCommitments": 15,
      "overallocationRisk": "HIGH"
    }
  ],
  "commitments": {
    "activeReleases": 12,
    "plannedReleases": 8,
    "totalFeatures": 234,
    "estimatedEffort": 1580.5
  },
  "overallocationWarnings": [
    {
      "owner": "jane.smith",
      "severity": "HIGH",
      "overallocationPercentage": 120.0
    }
  ]
}
```

## Service Layer
- Implement `DashboardService` for dashboard data aggregation and calculations
- Implement `PlanningAnalyticsService` for analytics calculations and insights

## Test Coverage
- Integration tests for all dashboard API endpoints
- Integration tests for all analytics API endpoints
- Tests for dashboard metric accuracy (completion rate, velocity, blocked time)
- Tests for timeline data inclusion in dashboard responses
- Tests for business logic validation (timeline adherence, risk level calculations)
- Tests for edge cases (empty releases, unassigned features, invalid data)
- Tests for data consistency and realistic value ranges
- Tests for error handling scenarios

## Acceptance Criteria
- Release dashboard provides comprehensive metrics and visualizations
- All required dashboard data is present and accurate
- Planning analytics deliver actionable insights
- Capacity planning correctly identifies overallocation risks
- Business logic rules are correctly implemented
- Edge cases are handled gracefully
- All calculations are mathematically correct
- Integration tests validate API behavior

## FAIL_TO_PASS
- `com.sivalabs.ft.features.api.controllers.DashboardControllerIntegrationTests`
- `com.sivalabs.ft.features.api.controllers.PlanningControllerIntegrationTests`

## PASS_TO_PASS
- `com.sivalabs.ft.features.api.controllers.FeatureControllerTests#shouldGetFeaturesByReleaseCode`

### CODE REVIEW

File: src/test/java/com/sivalabs/ft/features/api/controllers/DashboardControllerIntegrationTests.java [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+        Map<String, Object> overview = (Map<String, Object>) map.get("overview");
+        assertThat(((Number) overview.get("totalFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("completedFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("inProgressFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("blockedFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("pendingFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("completionPercentage")).doubleValue())
+                .isBetween(0.0, 100.0);
+
+        Map<String, Object> health = (Map<String, Object>) map.get("healthIndicators");
```
Comment (DmitriyChernyakSetronica) [inline src/test/java/com/sivalabs/ft/features/api/controllers/DashboardControllerIntegrationTests.java]: ```healthIndicators``` is missing from the requirements but is used in tests—this violates the black-box principle. The requirements need to be either amended or the test needs to be modified.

The same applies to the following fields. : 
```
Fields NOT in Requirements but Tested:

estimatedDaysRemaining - Not mentioned in requirements
percentageOfTime - Not mentioned in requirements
blockageReasons - Not mentioned in requirements
featuresPerWeek - Not mentioned in requirements
averageCycleTime - Not mentioned in requirements
utilizationRate - Not mentioned in requirements
```

File: src/test/java/com/sivalabs/ft/features/api/controllers/DashboardControllerIntegrationTests.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        assertThat(result).hasStatusOk();
+        Map<String, Object> map =
+                objectMapper.readValue(result.getResponse().getContentAsString(), new TypeReference<>() {});
+
+        assertThat(map.get("releaseCode")).isEqualTo("IDEA-2023.3.8");
+        assertThat(map.get("releaseStatus")).asString().isIn("DRAFT", "RELEASED");
+        assertThat(((Number) map.get("completionRate")).doubleValue()).isBetween(0.0, 100.0);
+
+        Map<String, Object> velocity = (Map<String, Object>) map.get("velocity");
+        assertThat(((Number) velocity.get("featuresPerWeek")).doubleValue()).isGreaterThanOrEqualTo(0.0);
```
Comment (DmitriyChernyakSetronica) [inline src/test/java/com/sivalabs/ft/features/api/controllers/DashboardControllerIntegrationTests.java]: // WEAK: Only checks >= 0, doesn't validate reasonableness
assertThat(((Number) velocity.get("featuresPerWeek")).doubleValue()).isGreaterThanOrEqualTo(0.0);

// BETTER: Should validate reasonable ranges
// featuresPerWeek should be < 50 (unrealistic to complete 50 features/week)
// averageCycleTime should be > 0.1 days (unrealistic to complete feature in 2.4 hours)

**Comment by DmitriyChernyakSetronica** [comment]
### 2. **Missing Edge Case Testing**

#### No Null Value Testing
```java
// ❌ MISSING: Tests don't verify null handling
// What if release has no features?
// What if all features are unassigned?
// What if release has no planned end date?
```

#### No Invalid Data Testing
```java
// ❌ MISSING: Tests don't verify data quality handling
// What if feature has invalid status?
// What if dates are in wrong order (end before start)?
// What if completion percentage calculation overflows?
```


#### Assumption About Test Data State
```java
// ❌ ASSUMPTION: Tests assume specific feature counts exist
assertThat(statusMap).containsKeys("DRAFT", "RELEASED");

// What if test database has no DRAFT releases?
// Tests should create known test data or handle empty states
```

File: src/test/java/com/sivalabs/ft/features/api/controllers/DashboardControllerIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        assertThat(((Number) overview.get("totalFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("completedFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("inProgressFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("blockedFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("pendingFeatures")).intValue()).isGreaterThanOrEqualTo(0);
+        assertThat(((Number) overview.get("completionPercentage")).doubleValue())
+                .isBetween(0.0, 100.0);
+
+        Map<String, Object> health = (Map<String, Object>) map.get("healthIndicators");
+        assertThat(health.get("timelineAdherence")).asString().isIn("ON_TRACK", "DELAYED", "AT_RISK");
```
Comment (DmitriyChernyakSetronica) [inline src/test/java/com/sivalabs/ft/features/api/controllers/DashboardControllerIntegrationTests.java]: ### 2. **Missing Business Logic Validation**

#### Timeline Adherence Logic
```java
//  WEAK: Only checks enum values, not logic
assertThat(health.get("timelineAdherence")).asString().isIn("ON_TRACK", "DELAYED", "AT_RISK");

//  BETTER: Validate business logic
if (release.plannedEndDate < currentDate && release.status != RELEASED) {
    assertThat(health.get("timelineAdherence")).isEqualTo("DELAYED");
}
```

**Comment by DmitriyChernyakSetronica** [comment]
### 1. **Business Logic Validation**

#### Missing Timeline Logic Tests
```java
// ❌ MISSING: No tests for timeline adherence calculation logic
@Test
void shouldMarkReleaseAsDelayedWhenPastPlannedDate() {
    // Test that overdue releases are marked as DELAYED
}

@Test
void shouldMarkReleaseAsAtRiskWhenApproachingDeadline() {
    // Test that releases near deadline with low completion are AT_RISK
}
```

#### Missing Risk Assessment Tests
```java
// ❌ MISSING: No tests for risk level calculation logic
@Test
void shouldCalculateHighRiskForManyBlockedFeatures() {
    // Test risk level increases with blocked feature percentage
}

@Test
void shouldCalculateLowRiskForOnTrackReleases() {
    // Test risk level is LOW for releases on track
}
```

### 2. **Data Quality Tests**

#### Missing Validation Tests
```java
// ❌ MISSING: No tests for data consistency
@Test
void shouldEnsureFeatureCountsAddUpCorrectly() {
    // Verify completed + inProgress + blocked + pending = total
}

@Test
void shouldEnsurePercentagesAreReasonable() {
    // Verify completion percentage matches feature counts
    // Verify utilization rates are realistic
}
```

### 3. **Error Handling Tests**

#### Missing Error Scenarios
```java
// ❌ MISSING: Limited error testing
@Test
void shouldHandleCorruptedFeatureData() {
    // Test with features having invalid status
}

@Test
void shouldHandleInconsistentDateData() {
    // Test with end date before start date
}
```

### 4. **Add Performance and Load Tests**

### 5. **Add Comprehensive Edge Cases**

```java
@Test
void shouldHandleAllEdgeCases() {
    // Test empty release
    testDashboardForRelease(createEmptyRelease());
    
    // Test fully completed release
    testDashboardForRelease(createCompletedRelease());
    
    // Test release with only blocked features
    testDashboardForRelease(createBlockedRelease());
    
    // Test release with no assigned features
    testDashboardForRelease(createUnassignedRelease());
}
```

**Comment by DmitriyChernyakSetronica** [comment]
@codecharlan  I didn't see any changes in the commit message about this :  

healthIndicators is missing from the requirements but is used in tests—this violates the black-box principle. The requirements need to be either amended or the test needs to be modified.

The same applies to the following fields. :

Fields NOT in Requirements but Tested:

estimatedDaysRemaining - Not mentioned in requirements
percentageOfTime - Not mentioned in requirements
blockageReasons - Not mentioned in requirements
featuresPerWeek - Not mentioned in requirements
averageCycleTime - Not mentioned in requirements
utilizationRate - Not mentioned in requirements

```[Dashboard Metrics & Planning Analytics Implementation](https://github.com/dpaia/feature-service/pull/177/changes/f4f9fa53d98b917e27004c010aae73e6fadd7d0e) https://github.com/dpaia/feature-service/issues/159  
Objective
Implement API endpoints and calculation logic to support release dashboards and comprehensive planning analytics.

Requirements
Dashboard API:
- GET /api/releases/{code}/dashboard - Release dashboard with overview, feature breakdown, progress metrics, health indicators, timeline data
- GET /api/releases/{code}/metrics - Detailed metrics for analytics (completion rate, velocity, blocked time, workload distribution)

Analytics API:
- GET /api/planning/health - Planning health report (releases by status, overdue/at-risk counts, planning accuracy)
- GET /api/planning/trends - Trend data (releases completed over time, average duration, planning accuracy)
- GET /api/planning/capacity - Capacity planning data (workload by owner, commitments, overallocation warnings)

Service Layer:
- Implement DashboardService for dashboard data aggregation and calculations
- Implement PlanningAnalyticsService for analytics calculations and insights

Test Coverage
- Integration tests for all dashboard API endpoints
- Integration tests for all analytics API endpoints
- Tests for dashboard metric accuracy (completion rate, velocity, blocked time)
- Tests for timeline data inclusion in dashboard responses
- Tests for business logic validation (timeline adherence, risk level calculations)
- Tests for edge cases (releases with no features, all unassigned features, etc.)
- Tests for data consistency (feature counts add up, percentages are correct)
- Tests for realistic value ranges (velocity metrics within reasonable bounds)
- Tests for error handling (invalid release codes, missing data scenarios)

Acceptance Criteria
- Release dashboard provides comprehensive metrics and visualizations
- Dashboard includes all required data: timeline, health indicators, feature breakdown, workload distribution
- Dashboard endpoints return accurate metrics (completion rate, velocity, blocked time)
- Planning analytics API delivers actionable insights
- Health reports accurately show releases by status and at-risk counts
- Trend data provides meaningful historical analysis
- Capacity planning identifies over allocation warnings correctly
- Planning accuracy is correctly calculated and reported
- All business logic rules are correctly implemented (timeline adherence, risk levels)
- Edge cases are handled gracefully (empty releases, invalid data)
- Field calculations are mathematically correct (percentages, rates, sums)
- Integration tests validate API behavior without implementation dependencies

FAIL_TO_PASS:
com.sivalabs.ft.features.api.controllers.DashboardControllerIntegrationTests,com.sivalabs.ft.features.api.controllers.PlanningControllerIntegrationTests
PASS_TO_PASS: com.sivalabs.ft.features.api.controllers.FeatureControllerTests#shouldGetFeaturesByReleaseCode```
